crates/ra_project_model/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+mod cargo_workspace;
+mod sysroot;
+
+use std::path::{Path, PathBuf};
+
+use failure::bail;
+use rustc_hash::FxHashMap;
+
+use ra_db::{CrateGraph, FileId};
+use ra_vfs::Vfs;
+
+pub use crate::{
+    cargo_workspace::{CargoWorkspace, Package, Target, TargetKind},
+    sysroot::Sysroot,
+};
+
+// TODO use own error enum?
+pub type Result<T> = ::std::result::Result<T, ::failure::Error>;
+
+#[derive(Debug, Clone)]
+pub struct ProjectWorkspace {
+    pub cargo: CargoWorkspace,
+    pub sysroot: Sysroot,
+}
+
+impl ProjectWorkspace {
+    pub fn discover(path: &Path) -> Result<ProjectWorkspace> {
+        let cargo_toml = find_cargo_toml(path)?;
+        let cargo = CargoWorkspace::from_cargo_metadata(&cargo_toml)?;
+        let sysroot = Sysroot::discover(&cargo_toml)?;
+        let res = ProjectWorkspace { cargo, sysroot };
+        Ok(res)
+    }
+
+    pub fn to_crate_graph(&self, vfs: &mut Vfs) -> CrateGraph {
```

Comment:
Could we avoid `vfs` dependecy by passing an `&mut dyn FnMut(&Path) ->FileId`?

Reply:
Yeah, that works. It might be a good idea to make this a proper trait (e.g. `FileLoader`) later? I'll leave it like this for now though.